### PR TITLE
legacy: Builder refactorization (part 5...)

### PR DESCRIPTION
**Pull request by alessio-perugini**
## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [ ] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] `UPGRADING.md` has been updated with a migration guide (for breaking changes)
- [ ] `configuration.schema.json` updated if new parameters are added.

### CODE REVIEW

File: commands/compile/compile.go [hunk 1]
```diff
@@ -170,18 +170,14 @@ func Compile(ctx context.Context, req *rpc.CompileRequest, outStream, errStream
 		coreBuildCachePath = buildCachePath.Join("core")
 	}
 
-	requiredTools, err := pme.FindToolsRequiredForBuild(targetPlatform, buildPlatform)
-	if err != nil {
+	if _, err := pme.FindToolsRequiredForBuild(targetPlatform, buildPlatform); err != nil {
 		return nil, err
 	}
```
Comment (cmaglie) [inline commands/compile/compile.go:175]: Maybe this call can be completely removed

File: arduino/builder/preprocess_sketch.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+func (b *Builder) PreprocessSketch(includes paths.PathList, lineOffset int, onlyUpdateCompilationDatabase bool) error {
+	// In the future we might change the preprocessor
+	normalOutput, verboseOutput, err := preprocessor.PreprocessSketchWithCtags(
+		b.sketch, b.buildPath, includes, lineOffset,
+		b.buildProperties, onlyUpdateCompilationDatabase,
+	)
+	if b.logger.Verbose() {
+		b.logger.WriteStdout(verboseOutput)
+	}
+	b.logger.WriteStdout(normalOutput)
```
Comment (cmaglie) [inline arduino/builder/preprocess_sketch.go]: `normalOutput` should be printed instead of `verboseOutput` not together.

**Review by cmaglie — CHANGES_REQUESTED** [review]

File: arduino/builder/builder.go [hunk 3]
```diff
@@ -64,12 +70,20 @@ func NewBuilder(
...
+		return nil, fmt.Errorf("invalid build properties: %w", err)
 	}
+	buildProperties.Merge(customBuildProperties)
+
+	return &Builder{
+		sketch:                sk,
+		buildProperties:       buildProperties,
+		coreBuildCachePath:    coreBuildCachePath,
+		jobs:                  jobs,
+		customBuildProperties: append(requestBuildProperties, "build.warn_data_percentage=75"),
```
[the remark this answers is not in the document]
Comment (alessio-perugini) [reply arduino/builder/builder.go]: In a later commit I actually rewrote it in the same way I found it in the commands/compile:

https://github.com/arduino/arduino-cli/pull/2312/files/1239ab2c20a6fc4ab599a6b3b2e3ca207ffaa954#diff-1755d7f375279693c8b657d7ba180919b5c92c2a1baee7d273d65feb149a7d8bR143

```go
customBuildPropertiesArgs := append(requestBuildProperties, "build.warn_data_percentage=75")
...
customBuildProperties: customBuildPropertiesArgs,
```

File: commands/compile/compile.go (hunk 1 above)
Comment (alessio-perugini) [reply commands/compile/compile.go:175]: I thought the same, but underneath if it doesn't find a tool release it fails. For this reason, I kept it

File: arduino/builder/builder.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmaglie) [reply arduino/builder/builder.go]: ah I see 👍🏼

File: arduino/builder/preprocess_sketch.go (hunk 2 above)
Comment (alessio-perugini) [reply arduino/builder/preprocess_sketch.go]: Fixed in the latest PR: 99ccfb5d4f45bda1c27c4c4b08a262a7635f7297

File: arduino/builder/build_options_manager.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+	var prevOpts *properties.Map
+	if err := json.Unmarshal(buildOptionsJSONPrevious, &prevOpts); err != nil || prevOpts == nil {
+		m.builderLogger.Info(tr("%[1]s invalid, rebuilding all", "build.options.json"))
+		return wipe()
+	}
+
+	// If SketchLocation path is different but filename is the same, consider it equal
+	if filepath.Base(m.currentOptions.Get("sketchLocation")) == filepath.Base(prevOpts.Get("sketchLocation")) {
+		m.currentOptions.Remove("sketchLocation")
```
[the remark this answers is not in the document]
Comment (alessio-perugini) [reply arduino/builder/build_options_manager.go:159]: Fixed in the lastest PR: b63def015068075d603a9ce19ed088316e577f28

**Review by cmaglie — APPROVED** [review]

**Pull request merged**